webview2_win32/src/callback.rs

```diff
@@ -1,248 +1,391 @@
-        arg_1: Arg1::Input,
-        arg_2: Arg2::Input,
-    ) -> windows::HRESULT {
-        let interface = this as *mut Self;
-        match (*interface).completed() {
-            Some(completed) => completed(Arg1::convert(arg_1), Arg2::convert(arg_2)),
-            None => S_OK,
-        }
+impl<'a, I: 'a + Interface> InvokeArg<'a> for Option<I> {
+    type Input = &'a Self;
+
+    fn convert(input: &'a Self) -> <Self as ClosureArg>::Output {
+        input.clone()
     }
 }
 
-/// Generic closure signature for [`EventCallback`].
-pub type EventClosure<'a, Arg1, Arg2> = Box<
-    dyn 'a
-        + FnMut(
-            <Arg1 as ClosureArg<'a>>::Output,
-            <Arg2 as ClosureArg<'a>>::Output,
-        ) -> windows::HRESULT,
+/// Generic closure signature for [`completed_callback`].
+pub type CompletedClosure<Arg1, Arg2> = Box<
+    dyn FnOnce(<Arg1 as ClosureArg>::Output, <Arg2 as ClosureArg>::Output) -> ::windows::Result<()>,
 >;
 
-/// All of the async events in [`WebView2`] use a repeatable handler interface which implements
-/// a COM method matching [`EventCallback::Invoke`]. The [`EventCallback::Arg1::Input`] and
-/// [`EventCallback::Arg2::Input`] parameter types vary depending on the async operation.
-#[allow(non_snake_case)]
-pub trait EventCallback<'a, T, Arg1, Arg2>: CallbackInterface<'a, T>
-where
-    T: Callback<'a>,
-    Arg1: ClosureArg<'a>,
-    Arg2: ClosureArg<'a>,
-{
-    fn event(&mut self) -> &mut EventClosure<'a, Arg1, Arg2>;
-
-    unsafe extern "system" fn Invoke(
-        this: windows::RawPtr,
-        arg_1: Arg1::Input,
-        arg_2: Arg2::Input,
-    ) -> windows::HRESULT {
-        let interface = this as *mut Self;
-        ((*interface).event())(Arg1::convert(arg_1), Arg2::convert(arg_2))
-    }
-}
+/// Generic closure signature for [`event_callback`].
+pub type EventClosure<Arg1, Arg2> = Box<
+    dyn FnMut(<Arg1 as ClosureArg>::Output, <Arg2 as ClosureArg>::Output) -> windows::Result<()>,
+>;
 
 #[completed_callback]
 pub struct CreateCoreWebView2EnvironmentCompletedHandler(
-    WebView2::ICoreWebView2CreateCoreWebView2EnvironmentCompletedHandler,
-    ErrorCodeArg,
-    InterfaceArg<WebView2::ICoreWebView2Environment>,
+    ICoreWebView2CreateCoreWebView2EnvironmentCompletedHandler,
+    HRESULT,
+    Option<ICoreWebView2Environment>,
 );
 
 #[completed_callback]
 pub struct CreateCoreWebView2ControllerCompletedHandler(
-    WebView2::ICoreWebView2CreateCoreWebView2ControllerCompletedHandler,
-    ErrorCodeArg,
-    InterfaceArg<WebView2::ICoreWebView2Controller>,
+    ICoreWebView2CreateCoreWebView2ControllerCompletedHandler,
+    HRESULT,
+    Option<ICoreWebView2Controller>,
 );
 
 #[event_callback]
-pub struct WebMessageReceivedEventHandler(
-    WebView2::ICoreWebView2WebMessageReceivedEventHandler,
-    InterfaceArg<WebView2::ICoreWebView2>,
-    InterfaceArg<WebView2::ICoreWebView2WebMessageReceivedEventArgs>,
+pub struct NewBrowserVersionAvailableEventHandler(
```

Comment:
Can we perhaps contribute an attribute of some kind to the metadata so that the Windows crate can do this automatically as it does for WinRT delegates?

Reply:
Some parts of this might still exist for WebView2 specifically, such as the generation of `wait_for_async_operation` in `#[completed_callback]`, but I could imagine a lot more of this going away in the future. The argument conversions are mostly nice-to-have's for the consumer, although the lifetime annotations on `&Option<Interface>` were tricky, which is why I still needed the generic `InvokeArg<'a>` trait, and it needed to be separate from the `ClosureArg` trait to avoid bringing that into the closure type and making it a generic parameter of the `#[implement]`  struct itself (not yet supported).

Reply:
We'll keep chipping away at the problem. Thanks again for the help with this.